internal/services/admin/v1/server.go

```diff
@@ -0,0 +1,176 @@
+package v1
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	msgqueue "github.com/hatchet-dev/hatchet/internal/msgqueue/v1"
+	contracts "github.com/hatchet-dev/hatchet/internal/services/admin/contracts/v1"
+	tasktypes "github.com/hatchet-dev/hatchet/internal/services/shared/tasktypes/v1"
+	"github.com/hatchet-dev/hatchet/pkg/repository/postgres/dbsqlc"
+	"github.com/hatchet-dev/hatchet/pkg/repository/postgres/sqlchelpers"
+	v1 "github.com/hatchet-dev/hatchet/pkg/repository/v1"
+	"github.com/hatchet-dev/hatchet/pkg/repository/v1/sqlcv1"
+)
+
+func (a *AdminServiceImpl) CancelTasks(ctx context.Context, req *contracts.CancelTasksRequest) (*contracts.CancelTasksResponse, error) {
+	tenant := ctx.Value("tenant").(*dbsqlc.Tenant)
+
+	tasksToCancel := []v1.TaskIdRetryCount{}
+
+	if len(req.Tasks) == 0 {
+		return &contracts.CancelTasksResponse{}, nil
+	}
+
+	// for tasks passed in directly, make sure the tenant id and retry counts match
+	if len(req.Tasks) > 0 {
+		taskIds := make([]int64, len(req.Tasks))
+		retryCounts := make([]int32, len(req.Tasks))
+
+		for i, task := range req.Tasks {
+			taskIds[i] = task.TaskId
+			retryCounts[i] = task.RetryCount
+		}
+
+		tasks, err := a.repo.Tasks().ListTasks(ctx, sqlchelpers.UUIDToStr(tenant.ID), taskIds)
+
+		if err != nil {
+			return nil, err
+		}
+
+		for i, task := range tasks {
+			if task.RetryCount == retryCounts[i] {
```

Comment:
nit: can we add a comment here? on initial read, I was wondering if comparing the `task.RetryCount` to `retryCounts[i]` is safe to do, since it seems like it'll depend on the ordering of the result of `ListTasks`